Display QMessageBox Dialogs For Errors

TheKoopaKingdom opened the pull request:
As a forum/Discord moderator, we see many people ask for support when their solution is simply to dump their system files and shared font. So, what this PR does is it displays a QMessageBox dialog when one of these errors happens.

A dialog box will be shown when the following happens:
 - The shared font is missing.
 - The system/home files are missing.
 - An encrypted ROM is loaded (This was already in, but I had to fix it because something was breaking it from being shown.).
 - OpenGL requirements not being met (Again, this was already in, but I'm not sure if it was really working.).
 - Any other unknown fatal error.

Here's how it works:
 - When the error happens within core, Core::System::GetInstance().SetStatus() is called.
 - In the main loop within bootmanager.cpp,  the Core Loop returns the status. It checks if it was successful or not, and if the latter, it emits the ErrorThrown signal.
 - At this point ErrorThrown has already been connected to the OnCoreError slot.
 - OnCoreError shuts down the game, checks the error type, and shows the dialog acoordingly.

With this system in place, adding new error messages is pretty easy, add a new ResultStatus entry, set it somewhere within core, and add the handler to OnCoreError.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/citra-emu/citra/2611)
<!-- Reviewable:end -->

wwylele commented on src/core/loader/loader.h:
```diff
@@ -100,11 +100,13 @@ class AppLoader : NonCopyable {
+     * @ return Result of operation.
      */
-    virtual boost::optional<u32> LoadKernelSystemMode() {
+    virtual ResultStatus LoadKernelSystemMode(boost::optional<u32>* system_mode) {
```
If you are outputting `system_mode` via the param, it should be just a raw pointer/reference.

Eventually we should change the return type of this function (and other functions in similar situation) to `variant<ResultStatus, u32>`. Can be in another PR, though

wwylele commented on src/core/hle/service/fs/fs_user.cpp:
```diff
@@ -128,6 +128,11 @@ static void OpenFileDirectly(Service::Interface* self) {
+        if (static_cast<u32>(archive_id) == 0x2345678A) {
+            LOG_ERROR(Service_FS, "oh noes");
+            Core::System::GetInstance().throwErr(-1);
+            return;
```
You should do either of the two here:
 - immediately stop the core, or
 - assign proper error code to `cmd_buff` so that the core won't run with unexpected response.

wwylele commented on src/core/hle/service/fs/fs_user.cpp:
```diff
@@ -128,6 +128,11 @@ static void OpenFileDirectly(Service::Interface* self) {
 
     ResultVal<ArchiveHandle> archive_handle = OpenArchive(archive_id, archive_path);
     if (archive_handle.Failed()) {
+        if (static_cast<u32>(archive_id) == 0x2345678A) {
```
Use `ArchiveIdCode::NCCH`. Also a note: as already suggested in the name, this archive is not only for system archive, but for general archive (including installed games etc.). For 99% of games I believe they only uses this for system archive, so this condition is fine, but don't forget suspicious cases like this: #2398

JayFoxRox commented on src/core/hle/service/fs/fs_user.cpp:
```diff
@@ -131,6 +132,10 @@ static void OpenFileDirectly(Service::Interface* self) {
         LOG_ERROR(Service_FS,
                   "failed to get a handle for archive archive_id=0x%08X archive_path=%s",
                   static_cast<u32>(archive_id), archive_path.DebugStr().c_str());
+        if (static_cast<u32>(archive_id) == 0x2345678A) {
```
What is this magic? Could it be a named constant maybe?

TheKoopaKingdom replied:
In my testing, games that required the system files always failed with that archive handle. I intended on waiting for input from a reviewer familiar with that, not sure exactly what that is/means.

wwylele replied:
Use `ArchiveIdCode::NCCH`. Also a note: as already suggested in the name, this archive is not only for system archive, but for general archive (including installed games etc.). For 99% of games I believe they only uses this for system archive, so this condition is fine, but don't forget suspicious cases like this: #2398

JayFoxRox commented on src/citra_qt/main.cpp:
```diff
@@ -646,14 +654,57 @@ void GMainWindow::UpdateStatusBar() {
-    return answer != QMessageBox::No;
+    return QMessageBox::question(this, tr("Citra"), tr("Are you sure you want to close Citra?"),
+                                 QMessageBox::Yes | QMessageBox::No,
+                                 QMessageBox::No) != QMessageBox::No;
```
This is harder to read than the previous code imo.

JayFoxRox commented on src/core/loader/loader.h:
```diff
@@ -100,11 +100,13 @@ class AppLoader : NonCopyable {
         // 96MB allocated to the application.
-        return 2;
+        system_mode = 2;
+        return ResultStatus::Success;
```
nit: I believe it's better to return a pair here instead of modifying `system_mode` by reference (where it's easy to miss what's going on).

wwylele replied:
Also, if you are outputting `system_mode` via the param, it should be just a raw pointer/reference.

Eventually we should change the return type of this function (and other functions in similar situation) to `variant<ResultStatus, u32>`. Can be in another PR, though

TheKoopaKingdom replied:
Alright, for now I'll use std::pair, not sure if boost::variant would be better, but the header failed to compile when I included it.

JayFoxRox commented on src/citra_qt/main.cpp:
```diff
@@ -646,14 +654,57 @@ void GMainWindow::UpdateStatusBar() {
+    case Core::System::ResultStatus::ErrorUnknown:
+        QMessageBox::critical(
+            this, "Fatal Error",
+            "Citra has encountered a fatal error, please see the log for more details.");
```
I believe we have an article (wiki or forums?) about dumping logs

JayFoxRox commented on src/citra_qt/main.cpp:
```diff
@@ -341,27 +340,33 @@ bool GMainWindow::LoadROM(const QString& filename) {
+        case Core::System::ResultStatus::ErrorOpenGL:
+            QMessageBox::critical(this, tr("Error while loading OpenGL!"),
+                                  tr("Your GPU may not support OpenGL 3.3, or you do not "
+                                     "have the latest graphics driver."));
```
Can be stricter imo. Something about "Citra not working until GPU driver or hardware is upgraded."

JayFoxRox commented on src/citra_qt/main.cpp:
```diff
@@ -341,27 +340,33 @@ bool GMainWindow::LoadROM(const QString& filename) {
+                   "Cartridges</a></li>"
+                   "<li><a href='https://citra-emu.org/wiki/Dumping-Installed-Titles/'>Dumping "
+                   "Installed Titles</a></li>"
+                   "</ul>"));
```
Can mention that a 3DS is required in the process so we prevent idiots from asking questions.

TheKoopaKingdom replied:
(This has been addressed.)

JayFoxRox commented on src/citra_qt/main.cpp:
```diff
@@ -315,8 +315,7 @@ bool GMainWindow::LoadROM(const QString& filename) {
-                              tr("Failed to initialize the video core!\n\n"
-                                 "Please ensure that your GPU supports OpenGL 3.3 and that you "
+                              tr("Your GPU may not support OpenGL 3.3, or you do not"
                                  "have the latest graphics driver."));
```
Could be more precise: "Could not initialize OpenGL 3.3 Core." or something.
(Mostly so we don't have to check what went wrong when someone asks for that error message)

JayFoxRox commented on src/core/core.cpp:
```diff
@@ -139,7 +151,7 @@ System::ResultStatus System::Init(EmuWindow* emu_window, u32 system_mode) {
 
     if (!VideoCore::Init(emu_window)) {
-        return ResultStatus::ErrorVideoCore;
+        return ResultStatus::ErrorOpenGL;
```
The old constant made more sense: Video Core not being able to init must not be related to OpenGL at all.

JayFoxRox commented on src/citra_qt/main.cpp:
```diff
@@ -335,7 +335,7 @@ bool GMainWindow::LoadROM(const QString& filename) {
 
         case Core::System::ResultStatus::ErrorSystemMode:
-            LOG_CRITICAL(Frontend, "Failed to load ROM!");
+            LOG_CRITICAL(Frontend, "Failed to load ROM!!!");
```
Not really necessary + doesn't fit to the rest of this commit.

JayFoxRox commented on src/citra_qt/bootmanager.cpp:
```diff
@@ -37,9 +37,17 @@ void EmuThread::run() {
+                // TODO: there's probably a better way to wait for the dialog to be closed
+                while (!ready);
+                if (stop_run)
+                    break;
```
yes, there is a better way and this could isn't going to fly.

TheKoopaKingdom replied:
This was from before Jrowe looked at the code for me, it's already much more improved in a later commit.

JayFoxRox commented on src/citra_qt/bootmanager.cpp:
```diff
@@ -37,9 +37,17 @@ void EmuThread::run() {
+                    break;
+
+                // Reset status to success so we dont run into the same error every time
+                Core::System::GetInstance().SetStatus(Core::System::ResultStatus::Success);
```
Shouldn't this happen at the start [or every error-free exit] of `RunLoop` instead?

JayFoxRox commented on src/citra_qt/bootmanager.h:
```diff
@@ -4,11 +4,11 @@
 
 #pragma once
 
-#include <atomic>
```
nit: Why was this removed? std::atomic is still used below

JayFoxRox commented on src/citra_qt/main.cpp:
```diff
@@ -646,14 +655,57 @@ void GMainWindow::UpdateStatusBar() {
 
+void GMainWindow::OnCoreError(Core::System::ResultStatus result) {
+    // Waiting for the dialog to be closed before shutting down causes a segfault, maybe because of
+    // the profiler
```
This feels like it should be a github issue, not just a comment somewhere in the code?

JayFoxRox commented on src/citra_qt/main.cpp:
```diff
@@ -646,14 +655,57 @@ void GMainWindow::UpdateStatusBar() {
+        break;
+
+    default:
+        break;
```
With the ability to add new Errors now, shouldn't this be handled similar to `ErrorUnknown`?

RavenHome1 commented:
closes https://github.com/citra-emu/citra/issues/2456

wwylele commented on src/core/loader/loader.h:
```diff
@@ -100,12 +100,12 @@ class AppLoader : NonCopyable {
      * @ return Result of operation.
      */
-    virtual ResultStatus LoadKernelSystemMode(boost::optional<u32>* system_mode) {
+    virtual ResultStatus LoadKernelSystemMode(boost::optional<u32>& system_mode) {
```
If you are outputting `system_mode` via the param, it should be just a raw pointer/reference.

Eventually we should change the return type of this function (and other functions in similar situation) to `variant<ResultStatus, u32>`. Can be in another PR, though

wwylele commented on src/core/hle/service/fs/fs_user.cpp:
```diff
@@ -131,6 +132,10 @@ static void OpenFileDirectly(Service::Interface* self) {
                   static_cast<u32>(archive_id), archive_path.DebugStr().c_str());
+        if (static_cast<u32>(archive_id) == 0x2345678A) {
+            Core::System::GetInstance().SetStatus(Core::System::ResultStatus::ErrorSystemFiles);
+            return;
```
You should do either of the two here:
 - immediately terminate the core, or
 - assign proper error code to `cmd_buff` so that the core won't run with garbage response.

jroweboy commented:
@RavenHome1 I suggest opening another issue about tilt controls. I agree that it is a recurring problem and it'd be nice to inform the user somehow, but thats way out of scope for this PR.

JayFoxRox commented on src/citra_qt/main.cpp:
```diff
@@ -341,27 +340,34 @@ bool GMainWindow::LoadROM(const QString& filename) {
+        case Core::System::ResultStatus::ErrorVideoCore:
+            QMessageBox::critical(this, tr("Error while initializing OpenGL 3.3 Core!"),
+                                  tr("Your GPU may not support OpenGL 3.3, or you do not "
+                                     "have the latest graphics driver."));
```
This is still misleading.

Reminder: VideoCore is just the name for our graphics emulation portion. It does not necessarily use OpenGL 3.3 Core (especially in the future).
There may be other errors in the future by the VideoCore subsystem which are not related to OpenGL.

All you can say at this point is: "An unknown error occured in the VideoCore subsystem" (which is probably not something we should say to users). It's probably best to suggest the user to ask for help and mention that the emulator reported an error in VideoCore.

(The other OpenGL error message in case of glad init failure is fine as it's strictly GL related)

uberhalit commented:
This breaks my Fire Emblem Fates and Awakening ROMs in BDE which both were working fine before 238.

jroweboy commented:
I've gotten several other reports from people that this caused issues so I'm going to remove this from bleeding edge for now. I couldn't cause the error locally, or I would have looked deeper into this problem. (If I had to take a guess, I wonder if some of the errors were previously non fatal for some games, and this PR makes them fatal errors.)

wwylele commented:
Tested some of my games and found the following issues:

 - Kirby Planet Robobot asking for a non-critical system archives: `failed to get a handle for archive archive_id=0x2345678A archive_path=[Binary: 020601009b00040000ffffff04000000]` and the dialog pops up. According to 3dbrew, this contains "Non-Nintendo TLS Root-CA Certificates". This file is currently not dumped by our archive dumper 3dsutil (can be easily added, though), and missing of this file doesn't prevent the game from running.

 - Some games may call `ThrowFatelError` on non-critical situation. For example, Pokemon asks for config block 0x00090000, which is just added in citra recently. If the game uses citra with save data inherited from an old version, the block doesn't exist and the game will `ThrowFatelError`, which doesn't prevent the game from continuing running before this PR. I am not sure about this... `ThrowFatelError` is still something that is not supposed happen in "normal" cases, but it can happen more frequently than we think.

TheKoopaKingdom commented:
@uberhalit What exactly happened when you tried running it?

(Will look into the false positive issues tomorrow.)

JayFoxRox commented on src/citra_qt/main.cpp:
```diff
@@ -663,8 +663,6 @@ void GMainWindow::UpdateStatusBar() {
 
 void GMainWindow::OnCoreError(Core::System::ResultStatus result) {
-    if (emu_thread != nullptr)
-        ShutdownGame();
```
You'll probably want to reflect this in the error messages. Otherwise we'll get support requests: "It said fatal error but then worked, is it safe to save?!".

Also I'm not sure if this might open a whole new can of worms: say an error in a tight loop - would that generate thousands of error messages?

---

I think a good way to solve this would be to utilize the new status bar for this and append warning messages if the game tried to do something like opening the font. This would signal that the emulation is tainted.
You could still show the error messages, but only the first time per run of a game?

(only my idea though and I feel there'll be different opinions about this)

TheKoopaKingdom replied:
>Also I'm not sure if this might open a whole new can of worms: say an error in a tight loop - would that generate thousands of error messages?

I had Core::RunLoop() reset the status every time.

>I think a good way to solve this would be to utilize the new status bar for this and append warning messages if the game tried to do something like opening the font. This would signal that the emulation is tainted.

Sounds like a good idea.

uberhalit commented:
@TheKoopaKingdom I get: 

> Citra was unable to locate the 3DS system archive.

wwylele commented:
What's the status of this PR? @TheKoopaKingdom are you still working on this?

wwylele commented on src/core/hle/service/fs/fs_user.cpp:
```diff
@@ -129,14 +129,61 @@ static void OpenFileDirectly(Service::Interface* self) {
+                        Core::System::ResultStatus::ErrorSystemFiles, "NG bad word list");
+                    return;
+                }
+            }
```
I feel like these checks should be placed [in `Archive_NCCH` handler, to which all 0x2345678A archives are dispatched](https://github.com/citra-emu/citra/blob/master/src/core/file_sys/archive_ncch.cpp#L40). Not sure if it is good to intrude so much, though.

TheKoopaKingdom replied:
I think I'll go with that, because that way, the error reporting will be covered by both OpenFileDirectly() and OpenArchive() (As in, the service function in FS_User), as opposed to just the former.

wwylele replied:
@TheKoopaKingdom you can still check the TID there. `data[0]` and `data[1]` are the TID

TheKoopaKingdom replied:
Added in [4c180e8](https://github.com/citra-emu/citra/pull/2611/commits/4c180e88904156620fe3545d9283999837c7c01b). Currently, situations in which there's a white-listed system file missing, there will be 2 messages logged: 
```
[   3.544855] Service.FS <Error> core/file_sys/archive_ncch.cpp:Open:68: Failed to get a handle for system data archive: NG bad word list. 
[   3.544870] Service.FS <Error> core/hle/service/fs/fs_user.cpp:OpenFileDirectly:136: Failed to get a handle for archive archive_id=0x2345678A archive_path=[Binary: 02030100db00040000ffffff01000000]
```
Not sure if this is too repetitive? I'm not really sure how to go about only having one message be printed, because of `Open()` returning a `ResultVal`, which `OpenFileDirectly()` can't really use to tell if the TID was white-listed or not.

wwylele replied:
I think two messages in log are fine, since log is meant to be verbose in order to report all status, and it is unnecessary to mess up with logic around log IMO.

TheKoopaKingdom commented:
Currently, aside from some commit squishing, this PR is about ready to be LGTM'd. The core functionality described in the PR description is still intact, but now has a few improvements:
 - The correct error code is assigned to core internally (Verified against HW.).
 - An option to quit back to the game list is provided within the message box.
 - The last error, if ignored by the user, is shown in the status bar.
 - Added a whitelist for the system title IDs that are dumped by 3dsutils to prevent dialogs being opened when they're unnecessary.

wwylele commented on src/core/core.cpp:
```diff
@@ -73,22 +77,32 @@ System::ResultStatus System::Load(EmuWindow* emu_window, const std::string& file
-        LOG_CRITICAL(Core, "Failed to determine system mode!");
-        return ResultStatus::ErrorSystemMode;
+    if (system_mode.second != Loader::ResultStatus::Success) {
+        LOG_CRITICAL(Core, "Failed to determine system mode (Error %i)!", system_mode.second);
```
`static_cast<int>(system_mode.second)`

wwylele commented on src/core/core.cpp:
```diff
@@ -3,6 +3,9 @@
 
 #include <memory>
+#include <utility>
+
```
no blank line between includes

wwylele commented on src/core/core.cpp:
```diff
@@ -73,22 +77,32 @@ System::ResultStatus System::Load(EmuWindow* emu_window, const std::string& file
-    const Loader::ResultStatus load_result{app_loader->Load()};
-    if (Loader::ResultStatus::Success != load_result) {
+    Loader::ResultStatus load_result = app_loader->Load();
+    if (load_result != Loader::ResultStatus::Success) {
```
What's the point of this change? I understand that `Loader::ResultStatus::Success != load_result` looks bad but... IMO the diff should be kept as clean as possible

wwylele commented on src/core/core.h:
```diff
@@ -6,6 +6,9 @@
 
 #include <memory>
 #include <string>
+
```
ditto, no blank line between includes

TheKoopaKingdom replied:
Aren't STD includes separated from library includes?

wwylele commented on src/core/core.h:
```diff
@@ -118,6 +140,9 @@ class System {
     static System s_instance;
+
+    ResultStatus status;
+    boost::optional<std::string> status_details;
```
why not just store this as plain `std::string` and represent "no details" as empty string, as `SetStatus` do?

wwylele commented on src/core/hle/service/fs/archive.cpp:
```diff
@@ -266,11 +266,9 @@ ResultVal<ArchiveHandle> OpenArchive(ArchiveIdCode id_code, FileSys::Path& archi
-    }
+    if (itr == id_code_map.end())
+        return ResultCode(ErrorDescription::FS_NotFound, ErrorModule::FS, ErrorSummary::NotFound,
+                          ErrorLevel::Status);
```
clean this diff

wwylele replied:
This can be replaced with the constant `ERROR_NOT_FOUND` defined in `file_sys/errors.h

wwylele commented on src/core/hle/service/fs/fs_user.cpp:
```diff
@@ -128,11 +128,12 @@ static void OpenFileDirectly(Service::Interface* self) {
+
+        LOG_ERROR(Service_FS,
+                  "Failed to get a handle for archive archive_id=0x%08X archive_path=%s",
+                  static_cast<u32>(archive_id), archive_path.DebugStr().c_str());
```
clean this diff

wwylele commented on src/core/loader/loader.h:
```diff
@@ -8,8 +8,11 @@
 #include <string>
+#include <utility>
 #include <vector>
+
```
no blank lines

wwylele commented on src/core/loader/loader.h:
```diff
@@ -100,11 +103,11 @@ class AppLoader : NonCopyable {
      * This function defaults to 2 (96MB allocated to the application) if it can't read the
      * information.
-     * @returns Optional with the kernel system mode
+     * @return A pair with the system mode (If found) and the result.
```
`@returns a pair of Optional with the kernel system mode and ResultStatus`

wwylele commented on src/core/loader/ncch.h:
```diff
@@ -179,9 +179,9 @@ class AppLoader_NCCH final : public AppLoader {
     /**
      * Loads the Exheader and returns the system mode for this application.
-     * @return Optional with the kernel system mode
+     * @return A pair with the system mode (If found) and the result.
```
ditto

chris062689 commented:
Has there been any comments on this since it hit Bleeding Edge? I'm kind of surprised we haven't seen any support topics asking for what the new alert messages mean.

bunnei commented:
I think this broke BE, untagging.

@TheKoopaKingdom can you address the above comments so that we can get this wrapped up?

wwylele commented:
Please resolve conflicts and squash commits

wwylele commented on src/core/file_sys/archive_ncch.cpp:
```diff
@@ -33,11 +34,45 @@ ArchiveFactory_NCCH::ArchiveFactory_NCCH(const std::string& nand_directory)
+                                                      "NG bad word list");
+            }
+        }
+        return ResultCode(ErrorDescription::FS_NotFound, ErrorModule::FS, ErrorSummary::NotFound,
```
This can be replaced with the constant ERROR_NOT_FOUND defined in `file_sys/errors.h

jroweboy commented on src/citra_qt/main.cpp:
```diff
@@ -328,27 +335,41 @@ bool GMainWindow::LoadROM(const QString& filename) {
+                   "Citra. A real 3DS is required.<br/><br/>"
+                   "For more information on dumping and decrypting games, please see the following "
+                   "wiki pages: <ul>"
+                   "<li><a href='https://citra-emu.org/wiki/Dumping-Game-Cartridges/'>Dumping Game "
```
This link is a 404

jroweboy commented on src/citra_qt/main.cpp:
```diff
@@ -328,27 +335,41 @@ bool GMainWindow::LoadROM(const QString& filename) {
+                   "wiki pages: <ul>"
+                   "<li><a href='https://citra-emu.org/wiki/Dumping-Game-Cartridges/'>Dumping Game "
+                   "Cartridges</a></li>"
+                   "<li><a href='https://citra-emu.org/wiki/Dumping-Installed-Titles/'>Dumping "
```
This link is a 404

jroweboy commented on src/citra_qt/main.cpp:
```diff
@@ -623,11 +649,67 @@ void GMainWindow::UpdateStatusBar() {
+        tr("The game you are trying to load requires additional files from your 3DS to be dumped "
+           "before playing.<br/><br/>For more information on dumping these files, please see the "
+           "following wiki page: <a "
+           "href='https://citra-emu.org/wiki/"
```
This link is a 404

TheKoopaKingdom replied:
Fixed in [dbbd3f6f0c7b62c107f9afc737e14279ab2241ea](https://github.com/citra-emu/citra/pull/2611/commits/dbbd3f6f0c7b62c107f9afc737e14279ab2241ea#diff-5e8bf22f5c3b3579c402308dd96b7563R660).

chris062689 commented:
This breakage was due to a recent update to the website that forces all wiki article names to lowercase.

chris062689 commented:
While we're editing the dialog box, can we modify the text at the bottom to say, in addition to "Would you like to quit back to the game list?" to additionally say "If you continue, your may experience bugs or corrupted save data.", basically an additional note that if they continue they're going to have a bad time...

TheKoopaKingdom commented:
> While we're editing the dialog box, can we modify the text at the bottom to say, in addition to "Would you like to quit back to the game list?" to additionally say "If you continue, your may experience bugs or corrupted save data.", basically an additional note that if they continue they're going to have a bad time...

@chris062689 Sure. Added in [dbbd3f6f0c7b62c107f9afc737e14279ab2241ea](https://github.com/citra-emu/citra/pull/2611/commits/dbbd3f6f0c7b62c107f9afc737e14279ab2241ea#diff-5e8bf22f5c3b3579c402308dd96b7563R662).

I said this on Discord, but it's worth mentioning that when a user does ignore the message box, a message is shown in the status bar, added in [63c870196ca8f81ec8acc823522e12a2ef0e5855](https://github.com/citra-emu/citra/pull/2611/commits/63c870196ca8f81ec8acc823522e12a2ef0e5855) and implemented in [32f5d4da9485ba68a23e52fcc6bef352c4e14a25](https://github.com/citra-emu/citra/pull/2611/commits/32f5d4da9485ba68a23e52fcc6bef352c4e14a25#diff-5e8bf22f5c3b3579c402308dd96b7563R704).

chris062689 commented:
LGTM

bunnei commented on src/core/file_sys/archive_ncch.cpp:
```diff
@@ -33,11 +35,44 @@ ArchiveFactory_NCCH::ArchiveFactory_NCCH(const std::string& nand_directory)
+                                                      "Mii data");
+            } else if (low == region_manifest) {
+                LOG_ERROR(Service_FS,
+                          "Failed to get a handle for shared data archive: region manifes");
```
nit: manifest?

bunnei commented on src/core/file_sys/archive_ncch.cpp:
```diff
@@ -33,11 +35,44 @@ ArchiveFactory_NCCH::ArchiveFactory_NCCH(const std::string& nand_directory)
     if (!file->IsOpen()) {
-        return ResultCode(-1); // TODO(Subv): Find the right error code
+        // High Title ID of the archive: The category (https://3dbrew.org/wiki/Title_list).
+        const u32 shared_data_archive = 0x0004009B;
```
These can all be constexpr

bunnei commented on src/core/loader/ncch.h:
```diff
@@ -179,9 +179,9 @@ class AppLoader_NCCH final : public AppLoader {
     /**
      * Loads the Exheader and returns the system mode for this application.
-     * @return Optional with the kernel system mode
+     * @returns a pair of Optional with the kernel system mode and ResultStatus
```
This comment reads weird now... (it kind of did before)

bunnei commented on src/core/core.cpp:
```diff
@@ -101,6 +113,8 @@ System::ResultStatus System::Load(EmuWindow* emu_window, const std::string& file
     }
+    // this->status will be used for errors while actually running the game
+    status = ResultStatus::Success;
     return ResultStatus::Success;
```
return status?

bunnei commented on src/core/loader/loader.h:
```diff
@@ -100,11 +101,11 @@ class AppLoader : NonCopyable {
      * This function defaults to 2 (96MB allocated to the application) if it can't read the
      * information.
-     * @returns Optional with the kernel system mode
+     * @returns a pair of Optional with the kernel system mode and ResultStatus.
```
Ditto with weird comment

bunnei commented on src/core/core.h:
```diff
@@ -105,6 +108,19 @@ class System {
+        status_details = details;
+    }
+
+    std::string GetStatusDetails() {
```
This can return `const std::string&`. Should also be a `const` method. E.g.
`const std::string GetStatusDetails)() const`

bunnei commented on src/core/core.h:
```diff
@@ -105,6 +108,19 @@ class System {
+        return status;
+    }
+
+    void SetStatus(ResultStatus new_status, std::string details = std::string()) {
```
The std::string arg here is not particularly performant. If we really need this, consider passing this by const reference (e.g. `const std::string&` and making it required. As an alternative, this can also just be a C-string if you need it to be optional, e.g. `const char* details = nullptr`, then `if (details) status_details = details` - prevents unnecessary construction/copies of std::string

bunnei commented on src/core/core.h:
```diff
@@ -105,6 +108,19 @@ class System {
     PerfStats perf_stats;
     FrameLimiter frame_limiter;
 
+    ResultStatus GetStatus() {
```
Should be a `const` method

bunnei commented on src/core/core.h:
```diff
@@ -130,6 +146,9 @@ class System {
 
     static System s_instance;
+
+    ResultStatus status;
```
Initialize this to something? (otherwise it will be garbage and can be hard to debug issues)

bunnei commented on src/core/core.cpp:
```diff
@@ -26,6 +26,7 @@ namespace Core {
 /*static*/ System System::s_instance;
 
 System::ResultStatus System::RunLoop(int tight_loop) {
+    this->status = ResultStatus::Success;
```
No need for `this->`

bunnei commented on src/citra_qt/main.cpp:
```diff
@@ -623,11 +649,69 @@ void GMainWindow::UpdateStatusBar() {
+    }
+
+    if (answer == QMessageBox::Yes) {
+        if (emu_thread != nullptr)
```
Can be just `if (emu_thread)`

bunnei commented on src/core/loader/ncch.cpp:
```diff
@@ -121,12 +121,16 @@ FileType AppLoader_NCCH::IdentifyType(FileUtil::IOFile& file) {
     if (!is_loaded) {
-        if (LoadExeFS() != ResultStatus::Success)
-            return boost::none;
+        ResultStatus res = LoadExeFS();
```
No need for this var, just call the load function in the if-statement?

TheKoopaKingdom replied:
`res` is used when returning the status on [line 128](https://github.com/citra-emu/citra/pull/2611/files#diff-97ddda51413fc4c2c643428d97920640R128).

bunnei commented on src/core/core.cpp:
```diff
@@ -73,22 +74,33 @@ System::ResultStatus System::Load(EmuWindow* emu_window, const std::string& file
 
-    const Loader::ResultStatus load_result{app_loader->Load()};
-    if (Loader::ResultStatus::Success != load_result) {
+    Loader::ResultStatus load_result = app_loader->Load();
```
No need for this var, just call the load function in the if-statement?

bunnei commented on src/citra_qt/bootmanager.cpp:
```diff
@@ -38,7 +38,10 @@ void EmuThread::run() {
                 emit DebugModeLeft();
 
-            Core::System::GetInstance().RunLoop();
+            Core::System::ResultStatus result = Core::System::GetInstance().RunLoop();
```
No need for this var, just call the function in the if-statement?

TheKoopaKingdom replied:
`result` is used when passing the status to `ErrorThrown` on [line 43](https://github.com/citra-emu/citra/pull/2611/files/dbbd3f6f0c7b62c107f9afc737e14279ab2241ea#diff-38ca3f21a73cc5c54e8fe31685b11829R43).

Alternatively, subbing something like `Core::System::GetInstance().GetStatus()` in for that parameter would be possible.

bunnei commented on src/citra_qt/main.cpp:
```diff
@@ -623,11 +649,69 @@ void GMainWindow::UpdateStatusBar() {
+    switch (result) {
+    case Core::System::ResultStatus::ErrorSystemFiles: {
+        QString message = "Citra was unable to locate a 3DS system archive";
+        if (details != std::string())
```
nit: Prefer `{` and `}` even on one-line if-statements. While pedantic, it can prevent some nasty white space bugs

bunnei replied:
(applies here and elsewhere in this change)
